File: database/impl/src/main/java/info/nightscout/database/impl/daos/workaround/TraceableDaoWorkaround.java
package info.nightscout.database.impl.daos.workaround;

import androidx.room.Transaction;

import app.aaps.database.entities.interfaces.TraceableDBEntry;
import info.nightscout.database.impl.daos.TraceableDao;
import info.nightscout.database.impl.daos.TraceableDaoKt;

// keep in java, it's easier
public interface TraceableDaoWorkaround<T extends TraceableDBEntry> {

    /**
     * Inserts a new entry
     *
     * @return The ID of the newly generated entry
     */
    @Transaction
    default long insertNewEntry(T entry) {
        return TraceableDaoKt.insertNewEntryImpl((TraceableDao<T>) this, entry);
    }

    /**
     * Updates an existing entry
     *
     * @return The ID of the newly generated HISTORIC entry
     */
    @Transaction
    default long updateExistingEntry(T entry) {
        return TraceableDaoKt.updateExistingEntryImpl((TraceableDao<T>) this, entry);
    }

}

File: database/impl/src/main/kotlin/app/aaps/database/daos/workaround/TraceableDaoWorkaround.kt
package app.aaps.database.daos.workaround

import androidx.room.Transaction
import app.aaps.database.daos.TraceableDao
import app.aaps.database.daos.insertNewEntryImpl
import app.aaps.database.daos.updateExistingEntryImpl
import app.aaps.database.entities.interfaces.TraceableDBEntry

interface TraceableDaoWorkaround<T : TraceableDBEntry> {

    /**
     * Inserts a new entry
     *
     * @return The ID of the newly generated entry
     */
    @Transaction
    fun insertNewEntry(entry: T): Long =
        (this as TraceableDao<T>).insertNewEntryImpl(entry)

    /**
     * Updates an existing entry
     *
     * @return The ID of the newly generated HISTORIC entry
     */
    @Transaction
    fun updateExistingEntry(entry: T): Long =
        (this as TraceableDao<T>).updateExistingEntryImpl(entry)
}
